[Ready to Review] Feature/registration endpoint

pattisdr opened the pull request:
# Purpose

Issue https://openscience.atlassian.net/browse/OSF-4844
Registrations have a different serializer than nodes, because registrations have more information than nodes. Currently, we are using the /nodes/ endpoint to serialize both registrations and plain nodes, which means registrations lose information. We should likely make a /registrations/ endpoint and pull registrations out of the /nodes/ endpoint. 

# Changes
- Adds endpoints `/registrations/` and `/registrations/{registration_id}/` which use RegistrationSerializer with type `registrations`
- Modifies `/nodes/{node_id}/registrations/` to use RegistrationSerializer.  Self link goes to registrations endpoint not nodes endpoint
- Retractions are not accessible under registrations endpoints

- Nodes List View and Nodes Detail view will still return registrations, using Node Serializer, with type `nodes`  

**Node Registrations List** `/v2/nodes/<n_id>/registrations/`
*uses RegistrationSerializer, type=`registrations`, and self links is `/v2/registrations/<reg_id>/`
![screen shot 2015-10-15 at 4 56 36 pm](https://cloud.githubusercontent.com/assets/9755598/10527284/a96f8b6c-735d-11e5-8aac-64b22cbdf38a.png)

**Registration List**
*uses RegistrationSerializer, type=`registrations`, and self links is `/v2/registrations/<reg_id>/`

**Registration Detail**
*uses RegistrationSerializer, type=`registrations`, and self links is `/v2/registrations/<reg_id>/`
[Excludes nodes that aren't registrations]

sloria commented on api/base/urls.py:
```diff
@@ -14,6 +14,7 @@
                          url(r'^$', views.root),
                          url(r'^applications/', include('api.applications.urls', namespace='applications')),
                          url(r'^nodes/', include('api.nodes.urls', namespace='nodes')),
+                         url(r'^registrations/', include('api.registrations.urls', namespace='registrations')),
```
Good call on creating a new app for registrations.

sloria commented on api/registrations/views.py:
```diff
@@ -0,0 +1,178 @@
+
+    The registration was initiated by this user.
+
+    ###
```
Remove extra lines.

sloria commented on api/registrations/views.py:
```diff
@@ -0,0 +1,178 @@
+
+    See the [JSON-API spec regarding pagination](http://jsonapi.org/format/1.0/#fetching-pagination).
+
+    ##Actions
```
Remove this header if there is no content for it.

sloria commented on api/registrations/views.py:
```diff
@@ -0,0 +1,178 @@
+        self:  the canonical api endpoint of this registration
+        html:  this registration's page on the OSF website
+
+    ##Actions
```
Remove this.

sloria commented on api/registrations/views.py:
```diff
@@ -0,0 +1,178 @@
+
+    serializer_class = RegistrationDetailSerializer
+
+    # overrides RetrieveUpdateDestroyAPIView
```
This should be `ListAPIView`.

sloria commented on api/nodes/views.py:
```diff
@@ -360,7 +364,10 @@ class NodeDetail(generics.RetrieveUpdateDestroyAPIView, NodeMixin):
     def get_object(self):
-        return self.get_node()
+        node = self.get_node()
+        if node.is_registration:
```
I'm not sure this is the desired behavior. A 404 isn't appropriate because the node does exist. However, I'm not sure that we should even return an error if requesting a registration from `/nodes/<id>/`. Registrations are nodes, after all. Thoughts, @brianjgeiger ?

brianjgeiger replied:
If they aren't part of the collection, then they return a 404 if you try to access them through that collection, despite how similar they seem.

sloria replied:
Discussed in person with Brian: let's allow registrations to be listed via the `/nodes/` endpoints for now. AFAIK, it is ok to have the same stored object have multiple representations. So if a registration is requested through a `/nodes/` URL, it's `self` link will be the `/nodes/<id>`. If requesting through `/registrations/` -> `/registrations/<id>`, etc.

abought replied:
The NodeList view [also excludes folders and deleted items](https://github.com/CenterForOpenScience/osf.io/pull/4444/files#diff-51eef6ea2a6c35e2b080c9a6fe591e38R220), but the NodeDetail view `get_object` seems to use slightly different rules. Just to check, is that intentional?

sloria commented on api/registrations/views.py:
```diff
@@ -0,0 +1,178 @@
+    def get_default_odm_query(self):
+        base_query = (
+            Q('is_deleted', 'ne', True) &
+            Q('is_registration', 'eq', True)
```
We should also exclude any retracted registrations by default.

sloria replied:
Disregard; as discussed in person, we shouldn't exclude retractions by default.

sloria commented:
Pass finiished.

pattisdr commented:
@sloria So I should be able to access a registration through a node detail view, but it won't show up in node list view?

sloria commented:
No, it should show in the node list view with the `nodes` representation, i.e. it's type will be `nodes` and will use the NodeSerializer.

pattisdr commented:
@sloria, https://github.com/CenterForOpenScience/osf.io/pull/4444#discussion-diff-42034615 you say I should also exclude retractions, which I took to mean, in addition to registrations.  

Is this what you want?
/v2/nodes/ shows all nodes and registrations, but excludes retractions.  Uses node serializer, type nodes.
/v2/registrations shows registrations, uses registration serializer type registrations

sloria commented:
I meant that retractions should be excluded from `/v2/registrations/`:

```
/v2/nodes/{node_id}registrations/  
/v2/registrations
```

Should have type `registrations` and be rendered with `NodeRegistrationSerializer`.

```
/v2/nodes/
```

Should include nodes + registrations, have type `nodes` and be rendered with `NodeSerializer`. 

Correct, @brianjgeiger ?

brianjgeiger commented:
Yes on everything that's not retractions-related. I haven't given any thought to retractions, but is there any case in which we should see them? Is the only path /v2/registrations/{registration_id}/ once it's been retracted?

sloria commented:
In cases where the client needs retractions, they should be able to `GET /v2/registrations/?retracted=true`.

pattisdr commented:
I don't know how to filter on that field. retracted, or is_retracted is not a field on the database.  I wasn't even sure how to include it in my query because it was getting ignored so I tacked it on at the end.

https://github.com/pattisdr/osf.io/blob/feature/registration_endpoint/api/registrations/views.py#L103

pattisdr commented:
I am removing filtering on retractions. /v2/registrations/ and /v2/nodes/{node_id}/registrations/ include all registrations, retracted included.  However, retracted is a field on the serializer.

sloria commented:
Looks good. CR1 passed.

abought commented on api/registrations/views.py:
```diff
@@ -0,0 +1,174 @@
+    )
+
+    required_read_scopes = [CoreScopes.NODE_BASE_READ]
+    required_write_scopes = [CoreScopes.NODE_BASE_WRITE]
```
Please create a new and separate item in `CoreScopes` to deal with registrations. There is a very reasonable use case in which someone might want to grant access to basic node metadata (the `NodeDetail` view) without necessarily letting a third party read or edit registration data.

abought commented on api/registrations/serializers.py:
```diff
@@ -0,0 +1,52 @@
+
+    # TODO: Finish me
+
+    # TODO: Override create?
```
Given the additions in this PR, and the note in update, this begs the question: do these TODO comments still need to be there?

pattisdr replied:
I think these notes should stay.

#TODO: finish me -  might refer to OSF-4436 https://openscience.atlassian.net/browse/OSF-4436. Looks like they want to add more fields to the serializer.

#TODO: Override create? might refer to the fact that we can't currently create registrations through the API.  This was something I had worked on in the summer but was put on hold to wait for some pre-reg decisions to be finalized.

sloria replied:
Correct on both comments, @pattisdr . Both these TODOs are still relevant.

abought commented on api/registrations/serializers.py:
```diff
@@ -0,0 +1,52 @@
+        return obj.absolute_url
+
+    def update(self, *args, **kwargs):
+        raise exceptions.ValidationError('Registrations cannot be modified.')
```
This serializer is currently only used in ReadOnly views; if they manage to call update something has gone horribly wrong. I suggest raising something more severe than a ValidationError- perhaps a blanket `APIException`?

abought commented:
Design question: I kind of agree with the original intent (to treat nodes and registrations separately), even if the PR explicitly went a different direction.

Advantage of the current approach:

1. Convenient for the dashboard view, where we are display both nodes and registrations and don't want to make a second request

Arguments against:

1. Multiple representations of data
2. Information leakage: In the current recommended set of scopes, access to Node Detail is handed out separately from access to Registration Data (the fact that you created a registration at all tells you something significant about the status of a project). Providing another path to serialize registrations reveals information we may not have intended to give out.
3. Intended vs unintended consequences: I suspect this originally happened because our Node class does many things at once.  Is our internal choice to lump data together justification for lumping these two things together externally?

To be fair: the fact that node serializer is missing many registration fields *limits* how much data is leaked, but the amount is still nonzero. 

"I love it" is a very different design rationale than "well, I guess".

abought commented on api/registrations/views.py:
```diff
@@ -0,0 +1,174 @@
+        base_permissions.TokenHasScope,
+    )
+
+    required_read_scopes = [CoreScopes.NODE_BASE_READ]
```
See comment below on scopes for registration views.

abought commented on api/registrations/views.py:
```diff
@@ -0,0 +1,174 @@
+        user = self.request.user
+        permission_query = Q('is_public', 'eq', True)
+        if not user.is_anonymous():
+            permission_query = (Q('is_public', 'eq', True) | Q('contributors', 'icontains', user._id))
```
Rather than overriding and copying the above, you should be able to build on the query (`permission_query |=` for short)

brianjgeiger commented:
My only immediate concern with data leakage is in the cases of embargoed or retracted registrations. Embargoes are probably fine as embargoes are based on permissions. Retracted we should filter out from this view.

Yes, future implementations would need to keep this in mind.

abought commented on api/registrations/views.py:
```diff
@@ -0,0 +1,174 @@
+    # overrides RetrieveAPIView
+    def get_serializer_context(self):
+        # Serializer needs the request in order to make an update to privacy
+        # TODO: The method it overrides already returns request (plus more stuff). Why does this method exist?
```
Take it as a given that I asked about this line.

Is there an answer?

pattisdr replied:
I don't see any reason to override get_serializer_context.  I had just blindly copied from NodeDetail.  Removing.  Also, in one of Sam's PR's, he is removing the one from NodeDetail.

abought commented on api/registrations/views.py:
```diff
@@ -0,0 +1,174 @@
+    def get_object(self):
+        registration = self.get_node()
+        if not registration.is_registration:
+            raise NotFound('This is not a registration.')
```
Not sure whether 400 or 404 is the better error code here: something *was* found, after all.

abought commented on tests/api_tests/registrations/views/test_registration_detail.py:
```diff
@@ -0,0 +1,70 @@
+        self.private_project = ProjectFactory(title="Project Two", is_public=False, creator=self.user)
+        self.public_registration = RegistrationFactory(project=self.public_project, creator=self.user)
+        self.private_registration = RegistrationFactory(project=self.private_project, creator=self.user)
+        self.public_project.save()
```
Are these save statements necessary (eg to create backrefs)?

pattisdr replied:
I think I can take them out.

abought commented on tests/api_tests/registrations/views/test_registration_detail.py:
```diff
@@ -0,0 +1,70 @@
+        url = '/{}registrations/{}/'.format(API_BASE, self.public_project._id)
+        res = self.app.get(url, auth=self.user.auth, expect_errors=True)
+        assert_equal(res.status_code, 404)
+        assert_equal(res.json['errors'][0]['detail'], "This is not a registration.")
```
Please make sure to add a new line at the end of files. (pycharm even has a setting to do this automatically)

abought commented:
CR2 pass complete.

Some small cleanup, as well as changing the scopes associated with these views. (this governs what information we give out)

Also the larger design question of whether we should be lumping nodes and registrations together this way- I don't view that as trivial, especially if it starts requiring multiple extra queryset filters maintained in two places.

If we separate the two, a test that the node endpoint doesn't give access to registration data would be a helpful addition.

pattisdr commented:
> My only immediate concern with data leakage is in the cases of embargoed or retracted registrations. 

Retractions are currently included in all the endpoints.  Nodes are excluded from /registrations.  But registrations are included in /nodes/.

1) `/v2/nodes/`  ..includes nodes, registrations, retractions
2) `/v2/nodes/{node_id}/registrations/` ..includes registrations, retractions
3) `/v2/registrations/` ..includes registrations, retractions

Talking with @sloria last week, retractions are included because you can access retractions in the UI.   @brianjgeiger, if retractions are removed from 1) will that satisfy this "data leakage"?  

Also, I don't know how to exclude retractions in my query because it is not a field on the database.  I can get the query, and then use list comprehensions to filter out retractions afterwards. But then this causes problems because we then have to do `ordering = ('-date_modified',)`

pattisdr commented:
After discussing with @sloria, @brianjgeiger, and @cwisecarver, we will exclude registrations from the nodes endpoint for now. At some point, when we have heterogeneous collections, we will allow returning both nodes and registrations but currently registrations will be excluded from nodes and nodes excluded from registrations.

abought commented:
CR2 complete. 

Overall the changes look good, but I have one question about a seeming different between list and detail view rules. Also, some conflicts to be resolved.

If we want we can include a note for QA to check filtering of the various flavors of registrations as features roll out (regular, embargoes, retracted, draft,  embargoed, etc).

cwisecarver commented:
If you follow the string get_object eventually does the same thing. https://github.com/CenterForOpenScience/osf.io/blob/develop/api/base/utils.py#L46

pattisdr commented:
>The NodeList view also excludes folders and deleted items, but the NodeDetail view get_object seems to use slightly different rules. Just to check, is that intentional?

Folders and deleted nodes are **already** excluded from the NodeDetail view, just in a different place.  Folders excluded in get_node https://github.com/CenterForOpenScience/osf.io/blob/develop/api/nodes/views.py#L57 and deleted nodes excluded in location above (thanks @cwisecarver)

abought commented:
> Folders and deleted nodes are already excluded from the NodeDetail view, just in a different place. Folders excluded in get_node https://github.com/CenterForOpenScience/osf.io/blob/develop/api/nodes/views.py#L57 and deleted nodes excluded in location above (thanks @cwisecarver)

Thanks for the find. I won't block on this PR based on that, then... but see below for notes going forward.

Honestly, I think the reason we missed this is because there are three separate methods responsible for controlling what the detail view returns: `get_object_or_error`, `NodeMixin.get_node`,  and `NodeDetail.get_object`. 

There's a good case for the first two functions (general fields common to all collections vs specific node behaviors)... but it'd be nice if the node-specific logic was all in one place, from a code maintainability  standpoint. Currently it's split across two functions/classes.

As we move forward and draw more distinctions between various kinds of "Nodes" (projects, registrations, collections), it may make sense to move more of that filtering into the queryset similar to how listview works. That will be easier to clean up once new endpoints are further developed.

abought commented:
Questions answered and conflicts resolved. CR2 pass complete.
